stability: photos ops/sec drops to and stays at 0

This is easy to reproduce locally. Run a single node cockroach cluster:

```
rm -fr cockroach-data; ./cockroach start
```

And then start up the `photos` load generator with 4 users:

```
photos --users 4
```

After about 10min, the ops/sec has dropped to 0 and no progress is being made. You can try to control-c the `photos` process which will initiate draining (letting existing operations finish) and witness that the ops never complete.

fwiw, one pair that I just looked at.

```
E160329 17:20:07.245822 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:07.246076 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Min rw=false pri=0.02253735 iso=SERIALIZABLE stat=PENDING epo=0 ts=1459286407.167567553,0 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.05319887 iso=SERIALIZABLE stat=PENDING epo=10 ts=1459286407.166200652,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min)
E160329 17:20:07.327555 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:07.327769 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Min rw=false pri=0.02253735 iso=SERIALIZABLE stat=PENDING epo=0 ts=1459286407.167567553,0 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.05319887 iso=SERIALIZABLE stat=PENDING epo=10 ts=1459286407.166200652,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min)
E160329 17:20:08.228959 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:08.229312 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.02253735 iso=SERIALIZABLE stat=PENDING epo=1 ts=1459286407.842119366,2 orig=1459286407.842119366,2 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.05319887 iso=SERIALIZABLE stat=PENDING epo=11 ts=1459286407.842119366,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:08.799815 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:08.800152 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.02253735 iso=SERIALIZABLE stat=PENDING epo=1 ts=1459286407.842119366,2 orig=1459286407.842119366,2 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.05319887 iso=SERIALIZABLE stat=PENDING epo=13 ts=1459286408.430900829,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:10.112338 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:10.112700 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.06790017 iso=SERIALIZABLE stat=PENDING epo=17 ts=1459286409.560830576,1 orig=1459286409.560830576,1 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.06790017 iso=SERIALIZABLE stat=PENDING epo=2 ts=1459286409.533906031,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min)
E160329 17:20:12.466505 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:12.466861 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=7 ts=1459286411.785389210,1 orig=1459286411.785389210,1 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=18 ts=1459286411.785389210,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:15.388416 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:15.388702 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=27 ts=1459286414.879976479,1 orig=1459286414.879976479,1 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=10 ts=1459286414.874308000,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min)
E160329 17:20:15.928572 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:15.928752 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=13 ts=1459286415.814312586,1 orig=1459286415.814312586,1 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=26 ts=1459286415.814312586,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min)
E160329 17:20:17.497598 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:17.497875 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=30 ts=1459286417.054332011,1 orig=1459286417.054332011,1 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=14 ts=1459286417.050497713,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min)
E160329 17:20:18.267424 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:18.267770 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=31 ts=1459286417.916403795,1 orig=1459286417.916403795,1 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=14 ts=1459286417.904881930,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min)
E160329 17:20:18.384986 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:18.385236 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=31 ts=1459286417.916403795,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=14 ts=1459286417.904881930,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:20:02 UTC 2016/"O\xfdQ[\xf9\x1b@W\xa9K\x82\xb3\a\xf6\xb5Q"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min)
E160329 17:20:22.660200 storage/replica_command.go:1120  FAILEDPUSH: 5f1a935e-9643-40a8-b386-9093c966bbb5 1dff26cf-0bb0-41b1-883d-96b5667a786f
E160329 17:20:22.660589 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=39 ts=1459286422.134642890,1 orig=1459286422.134642890,1 max=1459286402.695728712,0 wto=false failed to push "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=24 ts=1459286422.116333091,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:20:05 UTC 2016/"\x93\x9c\x86\xc1\xb4fJo\xa1\xcbnw\xc9k\xb9."/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:44 UTC 2016/"\x9db-\xf8K\x98J`\xa6I\x0f\x9at\xef\xcdj"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad",/Min)
E160329 17:20:24.733941 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:24.734116 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=30 ts=1459286424.603028160,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=42 ts=1459286424.463535193,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:25.013324 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:25.013606 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=30 ts=1459286424.603028160,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=42 ts=1459286424.463535193,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:25.848331 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:25.848986 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=31 ts=1459286425.721464840,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=43 ts=1459286425.527016811,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
E160329 17:20:26.082712 storage/replica_command.go:1120  FAILEDPUSH: 1dff26cf-0bb0-41b1-883d-96b5667a786f 5f1a935e-9643-40a8-b386-9093c966bbb5
E160329 17:20:26.083039 storage/store.go:1645  error: txn "sql/executor.go:380 sql txn" id=1dff26cf key=/Table/53/1/""/Tue Mar 29 21:19:42 UTC 2016/"\xe5a\xa9\x1e\x1f\rMը\x1bfv\xaf@\x82\xad" rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=31 ts=1459286425.721464840,1 orig=1459286407.167567553,0 max=1459286407.167567553,0 wto=false failed to push "sql/executor.go:380 sql txn" id=5f1a935e key=/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1 rw=true pri=0.08745822 iso=SERIALIZABLE stat=PENDING epo=43 ts=1459286425.527016811,1 orig=1459286402.695728712,0 max=1459286402.695728712,0 wto=false txn: <nil>, ba: PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min), PushTxn [/Table/53/1/""/Tue Mar 29 21:19:40 UTC 2016/"LE\x890'4C)\xa1\x86\xf4\x1b\x83\x10\xacQ"/4/1,/Min)
```

Note the identical priorities. If that corresponded to on-disk state, one of the pushes should have succeeded.

--- a/kv/txn_coord_sender.go
+++ b/kv/txn_coord_sender.go
@@ -440,7 +440,7 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*r
 	if tc.linearizable && sleepNS > 0 {
 		defer func() {
 			if log.V(1) {
-				log.Infof("%v: waiting %s on EndTransaction for linearizability", br.Txn.Short(), util.TruncateDuration(sleepNS, time.Millisecond))
+				log.Infof("%v: waiting %s on EndTransaction for linearizability", br.Txn.ID.Short(), util.TruncateDuration(sleepNS, time.Millisecond))
 			}
 			time.Sleep(sleepNS)
 		}()

--- a/roachpb/data.go
+++ b/roachpb/data.go
@@ -26,7 +26,6 @@ import (
 	"math/rand"
 	"sort"
 	"strconv"
-	"strings"
 	"sync"
 	"time"
 
@@ -615,15 +614,6 @@ func NewTransaction(name string, baseKey Key, userPriority UserPriority,
 	}
 }
 
-// GetMeta returns the Transaction's metadata, or nil if the transaction
-// is nil.
-func (t *Transaction) GetMeta() *TxnMeta {
-	if t == nil {
-		return nil
-	}
-	return &t.TxnMeta
-}
-
 // Clone creates a copy of the given transaction. The copy is "mostly" deep,
 // but does share pieces of memory with the original such as Key, ID and the
 // keys with the intent spans.
@@ -840,16 +830,8 @@ func (t Transaction) String() string {
 		fmt.Fprintf(&buf, "%q ", t.Name)
 	}
 	fmt.Fprintf(&buf, "id=%s key=%s rw=%t pri=%.8f iso=%s stat=%s epo=%d ts=%s orig=%s max=%s wto=%t",
-		t.Short(), t.Key, t.Writing, floatPri, t.Isolation, t.Status, t.Epoch, t.Timestamp, t.OrigTimestamp, t.MaxTimestamp, t.WriteTooOld)
+		t.ID.Short(), t.Key, t.Writing, floatPri, t.Isolation, t.Status, t.Epoch, t.Timestamp, t.OrigTimestamp, t.MaxTimestamp, t.WriteTooOld)
 	return buf.String()
-}
-
-// Short returns the short form of the Transaction's UUID.
-func (t Transaction) Short() string {
-	if t.ID == nil {
-		return strings.Repeat("?", 8)
-	}
-	return t.ID.String()[:8]
 }
 
 // ResetObservedTimestamps clears out all timestamps recorded from individual

--- a/storage/engine/mvcc.go
+++ b/storage/engine/mvcc.go
@@ -1047,7 +1047,13 @@ func mvccPutInternal(
 			return nil
 		}
 	}
-	buf.newMeta = MVCCMetadata{Txn: txn.GetMeta(), Timestamp: timestamp}
+	{
+		var txnMeta *roachpb.TxnMeta
+		if txn != nil {
+			txnMeta = &txn.TxnMeta
+		}
+		buf.newMeta = MVCCMetadata{Txn: txnMeta, Timestamp: timestamp}
+	}
 	newMeta := &buf.newMeta
 
 	versionKey := metaKey

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1605,7 +1605,7 @@ func (r *Replica) checkIfTxnAborted(b engine.Engine, txn roachpb.Transaction) *r
 	if aborted {
 		// We hit the cache, so let the transaction restart.
 		if log.V(1) {
-			log.Infof("found abort cache entry for %s with priority %d", txn.Short(), entry.Priority)
+			log.Infof("found abort cache entry for %s with priority %d", txn.ID.Short(), entry.Priority)
 		}
 		newTxn := txn.Clone()
 		if entry.Priority > newTxn.Priority {

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -992,6 +992,10 @@ func (r *Replica) PushTxn(
 ) (roachpb.PushTxnResponse, error) {
 	var reply roachpb.PushTxnResponse
 
+	if args.Now.Equal(roachpb.ZeroTimestamp) {
+		return reply, util.Errorf("the field Now must be provided")
+	}
+
 	if args.PusherTxn.ID != nil {
 		reply.Txn = &args.PusherTxn
 	}
@@ -1071,51 +1075,49 @@ func (r *Replica) PushTxn(
 		return reply, nil
 	}
 
-	// pusherWins bool is true in the event the pusher prevails.
-	var pusherWins bool
-
 	priority := args.PusherTxn.Priority
 
-	// Check for txn timeout.
-	if reply.PusheeTxn.LastHeartbeat == nil {
-		reply.PusheeTxn.LastHeartbeat = &reply.PusheeTxn.Timestamp
-	}
-	if args.Now.Equal(roachpb.ZeroTimestamp) {
-		return reply, util.Errorf("the field Now must be provided")
-	}
-	// Compute heartbeat expiration (all replicas must see the same result).
-	expiry := args.Now
-	expiry.WallTime -= 2 * DefaultHeartbeatInterval.Nanoseconds()
+	var pusherWins bool
+	var reason string
 
-	if reply.PusheeTxn.LastHeartbeat.Less(expiry) {
-		if log.V(1) {
-			log.Infof("pushing expired txn %s", reply.PusheeTxn)
-		}
-		pusherWins = true
+	lastActive := reply.PusheeTxn.OrigTimestamp
+	if realHB := reply.PusheeTxn.LastHeartbeat; realHB != nil {
+		lastActive.Forward(*realHB)
+	}
+	switch {
+	case lastActive.Less(args.Now.Add(-2*DefaultHeartbeatInterval.Nanoseconds(), 0)):
+		reason = "pushee is expired"
 		// When cleaning up, actually clean up (as opposed to simply pushing
 		// the garbage in the path of future writers).
 		args.PushType = roachpb.PUSH_ABORT
-	} else if reply.PusheeTxn.Isolation == roachpb.SNAPSHOT && args.PushType == roachpb.PUSH_TIMESTAMP {
-		if log.V(1) {
-			log.Infof("pushing timestamp for snapshot isolation txn")
-		}
 		pusherWins = true
-	} else if args.PushType == roachpb.PUSH_TOUCH {
+	case args.PushType == roachpb.PUSH_TOUCH:
 		// If just attempting to cleanup old or already-committed txns,
 		// pusher always fails.
 		pusherWins = false
-	} else if reply.PusheeTxn.Priority < priority ||
-		(reply.PusheeTxn.Priority == priority && args.PusherTxn.ID != nil &&
-			(args.PusherTxn.Timestamp.Less(reply.PusheeTxn.Timestamp) ||
-				(args.PusherTxn.Timestamp.Equal(reply.PusheeTxn.Timestamp) &&
-					bytes.Compare(reply.PusheeTxn.ID.GetBytes(), args.PusherTxn.ID.GetBytes()) < 0))) {
-		// Pusher wins based on priority; if priorities are equal, order
-		// by lower txn timestamp; if priorities & timestamps are equal,
-		// the greater transaction ID wins.
-		if log.V(1) {
-			log.Infof("pushing intent from txn with lower priority %s vs %d", reply.PusheeTxn, priority)
-		}
+	case args.PushType == roachpb.PUSH_TIMESTAMP &&
+		reply.PusheeTxn.Isolation == roachpb.SNAPSHOT:
+		// Can always push a SNAPSHOT txn's timestamp.
+		reason = "pushee is SNAPSHOT"
 		pusherWins = true
+	case reply.PusheeTxn.Priority != priority:
+		reason = "priority"
+		pusherWins = reply.PusheeTxn.Priority < priority
+	case args.PusherTxn.ID == nil:
+		reason = "equal priorities; pusher not transactional"
+		pusherWins = false
+	default:
+		reason = "equal priorities; greater ID wins"
+		pusherWins = bytes.Compare(reply.PusheeTxn.ID.GetBytes(),
+			args.PusherTxn.ID.GetBytes()) < 0
+	}
+
+	if log.V(1) && reason != "" {
+		s := "pushed"
+		if !pusherWins {
+			s = "failed to push"
+		}
+		log.Infof("%s "+s+" %s: %s", args.PusherTxn.ID.Short(), reply.PusheeTxn.ID.Short(), reason)
 	}
 
 	if !pusherWins {

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -533,7 +533,7 @@ func (r *Replica) EndTransaction(
 		var err error
 		if txnAutoGC && len(externalIntents) == 0 {
 			if log.V(1) {
-				log.Infof("auto-gc'ed %s (%d intents)", h.Txn.Short(), len(args.IntentSpans))
+				log.Infof("auto-gc'ed %s (%d intents)", h.Txn.ID.Short(), len(args.IntentSpans))
 			}
 			err = engine.MVCCDelete(batch, ms, key, roachpb.ZeroTimestamp, nil /* txn */)
 		} else {

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -2989,6 +2989,13 @@ func TestPushTxnHeartbeatTimeout(t *testing.T) {
 // priority are pushed; if priorities are equal, then the txns
 // are ordered by txn timestamp, with the more recent timestamp
 // being pushable.
+// TODO(tschottdorf): we should have a randomized version of this test which
+// also simulates the client proto and persisted record diverging. For example,
+// clients may be using a higher timestamp for their push or the persisted
+// record (which they are not using) might have a higher timestamp, and even
+// in the presence of such skewed information, conflicts between two (or more)
+// conflicting transactions must not deadlock (see #5685 for an example of this
+// happening with older code).
 func TestPushTxnPriorities(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	tc := testContext{}
@@ -3003,23 +3010,25 @@ func TestPushTxnPriorities(t *testing.T) {
 		pushType                       roachpb.PushTxnType
 		expSuccess                     bool
 	}{
-		// Pusher has higher priority succeeds.
+		// Pusher with higher priority succeeds.
+		{2, 1, ts1, ts1, roachpb.PUSH_TIMESTAMP, true},
 		{2, 1, ts1, ts1, roachpb.PUSH_ABORT, true},
-		// Pusher has lower priority fails.
+		// Pusher with lower priority fails.
 		{1, 2, ts1, ts1, roachpb.PUSH_ABORT, false},
 		{1, 2, ts1, ts1, roachpb.PUSH_TIMESTAMP, false},
-		// Pusher has lower priority fails, even with older txn timestamp.
+		// Pusher with lower priority fails, even with older txn timestamp.
 		{1, 2, ts1, ts2, roachpb.PUSH_ABORT, false},
-		// Pusher has lower priority, but older txn timestamp allows success if !abort.
+		// Pusher has lower priority, but older txn timestamp allows success if
+		// !abort since there's nothing to do.
 		{1, 2, ts1, ts2, roachpb.PUSH_TIMESTAMP, true},
-		// With same priorities, older txn timestamp succeeds.
-		{1, 1, ts1, ts2, roachpb.PUSH_ABORT, true},
-		// With same priorities, same txn timestamp fails (orders by txn ID).
+		// With same priorities, larger Txn ID wins. Timestamp does not matter
+		// (unless it implies that nothing needs to be pushed in the first
+		// place; see above).
+		// Note: in this test, the pusher has the larger ID.
 		{1, 1, ts1, ts1, roachpb.PUSH_ABORT, true},
 		{1, 1, ts1, ts1, roachpb.PUSH_TIMESTAMP, true},
-		// With same priorities, newer txn timestamp fails.
-		{1, 1, ts2, ts1, roachpb.PUSH_ABORT, false},
-		{1, 1, ts2, ts1, roachpb.PUSH_TIMESTAMP, false},
+		{1, 1, ts2, ts1, roachpb.PUSH_ABORT, true},
+		{1, 1, ts2, ts1, roachpb.PUSH_TIMESTAMP, true},
 		// When touching, priority never wins.
 		{2, 1, ts1, ts1, roachpb.PUSH_TOUCH, false},
 		{1, 2, ts1, ts1, roachpb.PUSH_TOUCH, false},

--- a/util/uuid/uuid.go
+++ b/util/uuid/uuid.go
@@ -31,6 +31,20 @@ type UUID struct {
 	uuid.UUID
 }
 
+// String returns the canonical representation of the UUID. For nil UUIDs,
+// returns xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.
+func (u *UUID) String() string {
+	if u == nil {
+		u = EmptyUUID
+	}
+	return u.UUID.String()
+}
+
+// Short returns the first eight characters of the output of String().
+func (u *UUID) Short() string {
+	return u.String()[:8]
+}
+
 // Bytes shadows (*github.com/satori/go.uuid.UUID).Bytes() to prevent confusing
 // our default proto stringer.
 // TODO(tschottdorf): fix upstream.

--- a/util/uuid/uuid.go
+++ b/util/uuid/uuid.go
@@ -32,16 +32,19 @@ type UUID struct {
 }
 
 // String returns the canonical representation of the UUID. For nil UUIDs,
-// returns xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.
+// returns <nil>.
 func (u *UUID) String() string {
 	if u == nil {
-		u = EmptyUUID
+		return "<nil>"
 	}
 	return u.UUID.String()
 }
 
 // Short returns the first eight characters of the output of String().
 func (u *UUID) Short() string {
+	if u == nil {
+		return u.String()
+	}
 	return u.String()[:8]
 }
 

